File: Semaphores/ConsumerProducer.c
/*
 
Do time analysis
 
Shared Buffer
1)Manage shared memory
Probably using Mutex
2)What happens if the buffer is full
Or multiple producers
So check if buffer is full
 
3) What if P>C
 Also check if the buffer is empty
 
 Also tweak around the number of consumers and Producers
*/
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <pthread.h>
#include <time.h>
#include <semaphore.h>
#include <unistd.h>
#define THREADS 2
pthread_mutex_t  mutexBuffer;
sem_t semAvailable;
sem_t semOccupied;
 
int buffer[20];
int count=0;
 
 // This step produces a
void *producer(void *args){
    // To get the essence of the consumer producer question we need a constant inflow of processes
    while(1){
        int a=rand() % 100;
        // Here we have to decrement the buffer by 1 and wait for an empty slot to add a  to the Buffer
        sem_wait(&semAvailable);
         // To avoid the race condition
        pthread_mutex_lock(&mutexBuffer);
       
        if (count<20){
             buffer[count] =a;
            // Add it to the buffer and  them increment the counter
            count++;
        }
        else{
            printf("Entities Skipped %d",a);
        }
       // Once a occupies a place in the buffer we need to increment the semaphore
        sem_post(&semOccupied);
        pthread_mutex_unlock(&mutexBuffer);
    }
 
}
 
void *consumer(void *args){
    int out=-1;
    // Here it is the inverse as we will wait for Sem Full and post on semAvailable
    // This is due to emptying of buffer in this case
    sem_wait(&semOccupied);
    pthread_mutex_lock(&mutexBuffer);
    // Here we have to read the value from the buffer and also check for empty buffer
    if (count>0){
        out=buffer[count-1];
        // Here we will be using STack so LIFO Policy Followed
        count--;
        // Decrement the count
    }
    pthread_mutex_unlock(&mutexBuffer);  
   
    // Print the element
    sem_post(&semAvailable);
    printf("Consumed %d\n",out);
 
}
 
 
int main(int argc, char *argv[]){
    srand(time(NULL));
    pthread_t threads[THREADS];
    // Initializing the Mutex
    pthread_mutex_init(&mutexBuffer,NULL);\
    // Initializing the semaphores
    sem_init(&semAvailable,0,20);
    sem_init(&semOccupied,0,0);
    // Creating two threads
    // So obviously we can create more consumer threads and producer threads
    for (int i=0;i<THREADS;i++){
        if (i%2==0){
           
            if (pthread_create(&threads[i],NULL,&producer,NULL)!=0){
                perror("Thread was not created");
            }
        }
        else{
            if (pthread_create(&threads[i],NULL,&consumer,NULL)!=0){
                 perror("Thread was not created");
            }
        }
        // Now we will join the threads together
        for (int j=0;j<THREADS;j++){
            if (pthread_join(threads[i],NULL)!=0){
                perror("Threads not joined");
            }
        }
    }
   
    pthread_mutex_destroy(&mutexBuffer,NULL);
    sem_destroy(&semAvailable);
    sem_destroy(&semOccupied);
 
    return 0;
 
 
 
 
 
}

File: Semaphores/ConsumerProducer.py
import random
import time
import os
import sys
import threading
import array

# Initializing the buffer size
BUFSIZE=5
# Initially the buffer has 0s
buf=[0]*BUFSIZE
# Defining the max Items to be produced in the buffer
max_items=5
# Buffer can be taken as a stack or a queue
# So we will initialize the producer and consumer index as same 0
p_index=0
c_index=0
counter=0
# Now we will need a Mutex for managing the critical section
# We will need two semaphores for managing the empty and occupied buffer
# Initially capacity available = BUFSIZE
available=threading.Semaphore(BUFSIZE)
# Initially space of Buffer occupied = 0
occupied=threading.Semaphore(0)
# Init mutex 
mutex = threading.Lock()
# For this we will have to use lock.acquire and lock.release
# Producer Class 
class Producer(threading.Thread):
    def run(self):
        global BUFSIZE,buf,p_index,c_index,counter
        global mutex,available,occupied
    # Producer Function provide a random Sleep time and a Random Entity to Produce
        i=0
        while(i<max_items):
            timesl =random.randrange(0,10,1)
            # # ranrange --> (start,stop,step)
            time.sleep(timesl/20)
            item=random.randrange(0,1000,1)
            # Generate random Number
            # Semaphore Waiting on available 
            available.acquire()
            # Locking mutex
            mutex.acquire()
            # Placing the Produced Item in the Buffer 
            if(counter<BUFSIZE):
                buf[counter]=item
                print("Entity Produced",item)
                p_index=p_index+1
                counter+=1
            else:
                # Checks and Balances
                print("Entity skipped")
            # Sem Post like on Operation on Occupied semaphore
            occupied.release()
            mutex.release()
            i+=1


# Consumer Thread Class
class Consumer(threading.Thread):
    def run(self):
        global BUFSIZE,buf,p_index,c_index,counter
        global mutex,available,occupied
        # Item to be consumed by the Buffer
        out=-1
        k=0
        while(k<max_items):
            timesl =random.randrange(0,10,1)
            # # ranrange --> (start,stop,step)
            time.sleep(timesl/20)
            occupied.acquire()
            mutex.acquire()
            if (counter>0):
                out=buf[counter-1]
                print("Entity Consumed" ,out)
                # Decrement the counter
                counter-=1

            mutex.release()
            available.release()
            k+=1

# Main Function
def main():
    print("Consumer Producer Question")
    threads=[]
    numthreads=2
    
    # Multi-Threading Approach can also be done if threads number increased
    # Initializing the Threads

    producer=[Producer() for i in range(numthreads)]
    consumer=[Consumer() for i in range(numthreads)]

    for i  in range(numthreads):
        if(i%2==0):
            producer[i].start()
            threads.append(producer[i])
           
        else:
            consumer[i].start()
            threads.append(consumer[i])
    # Join the threads
    for i in range(numthreads):
        threads[i].join()

    print("All Threads Joined")
    # Joining the Threads
    # for t in threads:
    #     t.join()
    # Starting Threads
    # for i in range(numthreads):
    #     producer[i].start()
    #     consumer[i].start()
    #     threads.append(producer[i])
    #     threads.append(consumer[i])
   
    # program exits
    print("--- %s seconds ---" % (time.time() - start_time))
    print("Exit Program")
    # After threads have been joined Terminate the Program
    sys.exit()

if __name__ == "__main__":
    start_time = time.time()
    main()
    




